### ISSUE TITLE
Uncaught errors due to rate limiting of the js.arcgis.com files

### ISSUE THREAD

**Issue by mrpintcom** [Esri/esri-loader#154]
### Expected behavior

If the user has been rate-limited from requesting more js.arcgis.com resource files on their localhost server, then show an appropriate error message or warning.

Project setup - webpack-dev-server, react, esri-loader v2.6.0 

### Actual behavior

If a developer refreshes the map frequently while making changes and testing, they can encounter an error originating in the dojo.js script used by the loader. Their cache must be disabled.

> Uncaught TypeError: Cannot read property 'insertBefore' of null
    at Function.u.injectUrl (dojo.js:24)
    at O (dojo.js:21)
    at a (dojo.js:4)
    at dojo.js:11
    at G (dojo.js:19)
    at sa (dojo.js:11)
    at dojo.js:22
    at a (dojo.js:4)
    at Oa (dojo.js:22)
    at q (dojo.js:21)
u.injectUrl @ dojo.js:24
O @ dojo.js:21
a @ dojo.js:4
(anonymous) @ dojo.js:11
G @ dojo.js:19
sa @ dojo.js:11
(anonymous) @ dojo.js:22
a @ dojo.js:4
Oa @ dojo.js:22
q @ dojo.js:21
(anonymous) @ dojo.js:23

If you wait a little while, the rate-limit will be reset and you can load the files again. 

### Steps to reproduce the behavior

Disable Cache in Chrome Developer Tools

Create a localhost server however you choose with a simple esri-loader implementation.

>          
       loadModules(["esri/Map",
            "esri/views/MapView",
            "esri/layers/WebTileLayer",
            "esri/Basemap",
            "esri/geometry/support/webMercatorUtils",
            "esri/geometry/Point",
            "esri/layers/GraphicsLayer",
        ], {}).then (([
               Map, MapView, WebTileLayer, Basemap, webMercatorUtils, Point, GraphicsLayer]
        ) => {
               console.log("MapView:`" + MapView) 

The rest of the code doesn't really matter because the error happens during the loading/fetching of the resource files.

Now, open the localhost server running your map, mine happened to be running at localhost:3000

Repeatedly refresh the map for a minute or so and you should start seeing the map failing to load with the error message above.

The `console.log("MapView", MapView)` in the code above will print out a valid function up until the point of failure. Then it started printing a "3" for me.

**Comment by mrpintcom** [comment]
At first I thought this was a localhost specific issue. But, with a live release, we are seeing it happen on a production URL as well.

**Comment by andygup** [comment]
@mrpintcom  Can you provide screenshots from the dev console of any network request/responses that are failing? 

I'm particularly interested in request/response headers and associated network error messages for failed requests.

**Comment by tomwayson** [comment]
@mrpintcom 

I like your suggestion that users should see "an appropriate error message or warning" when hitting rate limits.

However, I would suspect that if you made a jsbin w/ just the ArcGIS API and replaced the above `loadModules()` call w/ a call to `require()` and then "Repeatedly refresh the map for a minute or so" you'd see the same results, right? It's not the case that esri-loader is preventing the ArcGIS API from showing some useful error message, or suppressing such a message.

If that's true, I don't think we should do anything about it here. It'd be better to open an issue here: https://github.com/Esri/jsapi-resources/issues

**Comment by mrpintcom** [comment]
@andygup @tomwayson Thanks for your responses.

@andygup - there are no failed requests in the dev console, just a lack of resources being returned which causes the inevitable issue:
![FailedToLoadMore](https://user-images.githubusercontent.com/3252979/54706281-28ebb880-4b04-11e9-9627-22cf0bf97591.png)

Compared to this network traffic when successful:
![RemainingResourcesLoaded](https://user-images.githubusercontent.com/3252979/54706388-66504600-4b04-11e9-84b6-19cca773b1b2.png)

@tomwayson - Here's a screenshot from a similar type setup where MapView fails instead of esri-loader:
![AdditionalFailureLocation](https://user-images.githubusercontent.com/3252979/54706606-ee365000-4b04-11e9-99b3-66b2498bc2a3.png)

@tomwayson - I think that last screenshot proves that the issue comes from **jsapi-resources** like you stated.
I will open a new issue over there - https://github.com/Esri/jsapi-resources/issues

Thanks!

**Comment by andygup** [comment]
FWIW I wasn't able to reproduce which, by itself, is just one additional data point. The HTTP header you included above shows a valid `200` response request. If the CDN hosting provider was rate limiting there would be a header attribute indicating there was a problem. 

Do you know if you have any intermediary routers, firewalls, internet service providers that could have rate limits set on them?

**Comment by andygup** [comment]
Oh, sorry, I re-read that and saw the headers were from a successful request. My question is still the same...are there any intermediary points you are aware of that are rate limiting?

**Comment by tomwayson** [comment]
I'm going to close this for now as it does not appear to be esri-loader related.

**Comment by aspcanada** [comment]
@tomwayson, what issue did you open at https://github.com/Esri/jsapi-resources/issues ? Can you link it please?

**Comment by tomwayson** [comment]
@aspcanada I did not. It was @mrpintcom that said he was going to.

**Comment by ArkadiuszGorski** [comment]
Hi @tomwayson @mrpintcom did You came with some fix/solution or You know something about related tickets? I've got same issue.

**Comment by tomwayson** [comment]
@ArkadiuszGorski please open an issue at https://github.com/Esri/jsapi-resources/issues

**Comment by mrpintcom** [comment]
I have discovered the root cause of our issue.

Our MapView instance is stored in a "Data Store" to be shared across components. We also have this React HTML code - 
`<div className="map-container" id={this.props.id} ref={node => this.mapContainer = node}
`

The ref attribute can change, so later in our code we have - 
`map.view.container = this.mapContainer;` 
This makes sure the MapView is hooked up to the right container.

Unfortunately, we weren't updating the Data Store after changing the container. So later, the wrong container reference could be used (depending on order of  async requests finishing).

For us, it was intermittent because it depended on several async requests coming in a specific order to trigger it.

TLDR; the wrong scope of our map instance was saved in a data store and had the wrong container set for the MapView.

**Comment by vannizhang** [comment]
I keep getting the same issue intermittently (see screenshot below), but very interestingly I only get this error in Chrome, I can never reproduce it in Firefox, not even in the **Incognito mode** of Chrome...

Maybe it is an issue caused by how Chrome cache works? Here is the version of my Chrome: **Version 76.0.3809.100 (Official Build) (64-bit)**

![image](https://user-images.githubusercontent.com/3142684/63111228-14400f00-bf42-11e9-9add-e53270d5517f.png)

**Comment by andygup** [comment]
Thanks all. So far this doesn't appear to be AmazonS3 rate limiting and since it's only happening on Chrome that suspiciously points to some sort of a browser issue. Rate limiting from the CDN should, in theory, trigger `HTTP 503` errors. 

**If someone is able to use Fiddler, Charles or Wireshark to capture the before, during and after TCP/IP traffic we would greatly appreciate the stack dump.** 😄

**Comment by tomwayson** [comment]
i'm locking this issue in the hope that will encourage someone to finally open an issue over on jsapi-resources where it can the attention of the appropriate people.
